src/HttpRequest.h

```diff
@@ -146,13 +149,38 @@ class HttpRequest: public Http::Message
 
     int imslen;
 
-    Ip::Address client_addr;
+    /// mark this request as initiated by Squid (rather than received on a client connection)
+    void selfInitiated();
+
+    /// whether this request was initiated by Squid (rather than received on a client connection)
+    bool isSelfInitiated() const { return selfInitiated_; }
+
+    /// supply Downloader-specific settings
+    void prepareForDownloader(Downloader *);
+
+    /// specify addresses manually when lacking client connection
+    void prepareForConnectionlessProtocol(const Ip::Address &fromAddr, const Ip::Address &localAddr);
+
+    /// the remote address of the client connection
+    const Ip::Address& clientAddr() const;
+
+    /// the local address of the client connection
+    const Ip::Address& myAddr() const;
+
+    /// the client connection manager of the underlying transaction, if any
+    CbcPointer<ConnStateData> &clientConnectionManager();
+
+    /// the client connection of the underlying transaction, if any
+    Comm::ConnectionPointer clientConnection() const;
```

Comment:
Done. However, since MasterXaction::clientConnection_ should be nil for self-initiated requests, I think we should change the implementation, returning simply masterXaction->clientConnection().